File: frontend/src/services/unified_query.ts
import { axiosInstance } from './api';

/**
 * Database types supported by the unified SQL semantic layer
 */
export enum DatabaseType {
  PostgreSQL = 'postgresql',
  MySQL = 'mysql',
  Doris = 'doris',
  Druid = 'druid',
}

/**
 * Unified query request
 * Uses DataFusion SQL syntax that will be automatically translated to target database dialect
 */
export interface UnifiedQueryRequest {
  /** SQL query in DataFusion syntax */
  query: string;
  /** Target database type */
  database_type: DatabaseType;
  /** Query timeout in seconds (default: 30) */
  timeout_secs?: number;
  /** Whether to automatically apply LIMIT if not present (default: true) */
  apply_limit?: boolean;
  /** LIMIT value to apply (default: 1000) */
  limit_value?: number;
}

/**
 * Unified query response
 * Contains original query, translated query, and execution results
 */
export interface UnifiedQueryResponse {
  /** Original query in DataFusion SQL syntax */
  original_query: string;
  /** Translated query in target database dialect */
  translated_query: string;
  /** Target database type */
  database_type: DatabaseType;
  /** Query results as JSON objects */
  results: Record<string, any>[];
  /** Number of rows returned */
  row_count: number;
  /** Execution time in milliseconds */
  execution_time_ms: number;
  /** Whether LIMIT was automatically applied */
  limit_applied: boolean;
  /** Execution timestamp */
  executed_at: string;
}

/**
 * Unified Query Service
 *
 * Provides API methods for executing SQL queries using DataFusion's unified SQL syntax.
 * Queries are automatically translated to the target database's dialect.
 *
 * @example
 * ```typescript
 * // Execute a DataFusion SQL query against PostgreSQL
 * const response = await unifiedQueryService.executeUnifiedQuery(
 *   'conn-123',
 *   {
 *     query: "SELECT * FROM users WHERE created_at >= CURRENT_DATE - INTERVAL '7' DAY",
 *     database_type: DatabaseType.PostgreSQL,
 *     timeout_secs: 30,
 *     apply_limit: true,
 *     limit_value: 1000
 *   }
 * );
 *
 * console.log('Original:', response.original_query);
 * console.log('Translated:', response.translated_query);
 * console.log('Results:', response.results);
 * ```
 */
export const unifiedQueryService = {
  /**
   * Execute a unified SQL query using DataFusion semantic layer
   *
   * @param connectionId - Database connection ID
   * @param request - Unified query request with DataFusion SQL
   * @returns Unified query response with translation details and results
   *
   * @throws {Error} If database type doesn't match connection or query validation fails
   */
  async executeUnifiedQuery(
    connectionId: string,
    request: UnifiedQueryRequest
  ): Promise<UnifiedQueryResponse> {
    const response = await axiosInstance.post<UnifiedQueryResponse>(
      `/connections/${connectionId}/unified-query`,
      {
        query: request.query,
        database_type: request.database_type,
        timeout_secs: request.timeout_secs ?? 30,
        apply_limit: request.apply_limit ?? true,
        limit_value: request.limit_value ?? 1000,
      }
    );
    return response.data;
  },

  /**
   * Get database type from connection
   * Helper method to determine which DatabaseType to use for a connection
   *
   * @param connectionType - Connection type string from backend
   * @returns Corresponding DatabaseType enum value
   */
  getDatabaseType(connectionType: string): DatabaseType {
    const normalized = connectionType.toLowerCase();
    switch (normalized) {
      case 'postgresql':
      case 'postgres':
        return DatabaseType.PostgreSQL;
      case 'mysql':
      case 'mariadb':
        return DatabaseType.MySQL;
      case 'doris':
        return DatabaseType.Doris;
      case 'druid':
        return DatabaseType.Druid;
      default:
        throw new Error(`Unsupported database type: ${connectionType}`);
    }
  },

  /**
   * Check if unified query is supported for a database type
   *
   * @param databaseType - Database type to check
   * @returns true if unified queries are supported
   */
  isUnifiedQuerySupported(databaseType: DatabaseType): boolean {
    // PostgreSQL and MySQL fully supported
    // Doris and Druid support coming in Phase 5
    return databaseType === DatabaseType.PostgreSQL || databaseType === DatabaseType.MySQL;
  },

  /**
   * Get human-readable database type name
   *
   * @param databaseType - Database type enum value
   * @returns Display name for the database type
   */
  getDatabaseTypeName(databaseType: DatabaseType): string {
    switch (databaseType) {
      case DatabaseType.PostgreSQL:
        return 'PostgreSQL';
      case DatabaseType.MySQL:
        return 'MySQL';
      case DatabaseType.Doris:
        return 'Apache Doris';
      case DatabaseType.Druid:
        return 'Apache Druid';
      default:
        return databaseType;
    }
  },
};

File: backend/src/models/unified_query.rs
// Unified Query Request Model
//
// This model represents a query request that can be executed against any database type
// using DataFusion's unified SQL semantics. The query will be automatically translated
// to the target database's dialect.

use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};

/// Database type enumeration for unified query execution
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum DatabaseType {
    /// PostgreSQL database
    PostgreSQL,
    /// MySQL database
    MySQL,
    /// Apache Doris database
    Doris,
    /// Apache Druid database
    Druid,
}

impl DatabaseType {
    /// Get the string representation of the database type
    pub fn as_str(&self) -> &'static str {
        match self {
            DatabaseType::PostgreSQL => "postgresql",
            DatabaseType::MySQL => "mysql",
            DatabaseType::Doris => "doris",
            DatabaseType::Druid => "druid",
        }
    }

    /// Parse database type from string
    pub fn from_str(s: &str) -> Result<Self, String> {
        match s.to_lowercase().as_str() {
            "postgresql" | "postgres" => Ok(DatabaseType::PostgreSQL),
            "mysql" => Ok(DatabaseType::MySQL),
            "doris" => Ok(DatabaseType::Doris),
            "druid" => Ok(DatabaseType::Druid),
            _ => Err(format!("Unsupported database type: {}", s)),
        }
    }
}

/// Unified query request model
///
/// This represents a query that will be executed using DataFusion's unified SQL semantics.
/// The query is written in DataFusion SQL and will be automatically translated to the
/// target database's dialect.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedQueryRequest {
    /// The SQL query in DataFusion syntax
    pub query: String,

    /// The target database type for execution
    pub database_type: DatabaseType,

    /// Optional timeout in seconds (defaults to 30)
    #[serde(default = "default_timeout")]
    pub timeout_secs: u64,

    /// Whether to apply automatic LIMIT if not present (defaults to true)
    #[serde(default = "default_apply_limit")]
    pub apply_limit: bool,

    /// The LIMIT value to apply if apply_limit is true (defaults to 1000)
    #[serde(default = "default_limit_value")]
    pub limit_value: usize,
}

fn default_timeout() -> u64 {
    30
}

fn default_apply_limit() -> bool {
    true
}

fn default_limit_value() -> usize {
    1000
}

impl UnifiedQueryRequest {
    /// Create a new unified query request
    pub fn new(query: String, database_type: DatabaseType) -> Self {
        Self {
            query,
            database_type,
            timeout_secs: default_timeout(),
            apply_limit: default_apply_limit(),
            limit_value: default_limit_value(),
        }
    }

    /// Create a query request with custom timeout
    pub fn with_timeout(mut self, timeout_secs: u64) -> Self {
        self.timeout_secs = timeout_secs;
        self
    }

    /// Create a query request with custom limit settings
    pub fn with_limit(mut self, apply_limit: bool, limit_value: usize) -> Self {
        self.apply_limit = apply_limit;
        self.limit_value = limit_value;
        self
    }
}

/// Unified query response model
///
/// This represents the result of executing a unified SQL query.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedQueryResponse {
    /// The original query in DataFusion SQL syntax
    pub original_query: String,

    /// The translated query in the target database's dialect
    pub translated_query: String,

    /// The target database type
    pub database_type: DatabaseType,

    /// Query results as JSON objects
    pub results: Vec<serde_json::Value>,

    /// Number of rows returned
    pub row_count: usize,

    /// Execution time in milliseconds
    pub execution_time_ms: u128,

    /// Whether a LIMIT was automatically applied
    pub limit_applied: bool,

    /// Timestamp when the query was executed
    pub executed_at: DateTime<Utc>,
}

impl UnifiedQueryResponse {
    /// Create a new unified query response
    pub fn new(
        original_query: String,
        translated_query: String,
        database_type: DatabaseType,
        results: Vec<serde_json::Value>,
        execution_time_ms: u128,
        limit_applied: bool,
    ) -> Self {
        let row_count = results.len();

        Self {
            original_query,
            translated_query,
            database_type,
            results,
            row_count,
            execution_time_ms,
            limit_applied,
            executed_at: Utc::now(),
        }
    }
}

/// Simplified request for backward compatibility
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleUnifiedQueryRequest {
    /// The SQL query
    pub query: String,
}

impl From<SimpleUnifiedQueryRequest> for String {
    fn from(req: SimpleUnifiedQueryRequest) -> String {
        req.query
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_database_type_str_conversion() {
        assert_eq!(DatabaseType::PostgreSQL.as_str(), "postgresql");
        assert_eq!(DatabaseType::MySQL.as_str(), "mysql");
        assert_eq!(DatabaseType::Doris.as_str(), "doris");
        assert_eq!(DatabaseType::Druid.as_str(), "druid");
    }

    #[test]
    fn test_database_type_parsing() {
        assert_eq!(
            DatabaseType::from_str("postgresql").unwrap(),
            DatabaseType::PostgreSQL
        );
        assert_eq!(
            DatabaseType::from_str("postgres").unwrap(),
            DatabaseType::PostgreSQL
        );
        assert_eq!(
            DatabaseType::from_str("mysql").unwrap(),
            DatabaseType::MySQL
        );
        assert_eq!(
            DatabaseType::from_str("MySQL").unwrap(),
            DatabaseType::MySQL
        );

        assert!(DatabaseType::from_str("unknown").is_err());
    }

    #[test]
    fn test_unified_query_request_defaults() {
        let req = UnifiedQueryRequest::new(
            "SELECT * FROM users".to_string(),
            DatabaseType::PostgreSQL,
        );

        assert_eq!(req.timeout_secs, 30);
        assert!(req.apply_limit);
        assert_eq!(req.limit_value, 1000);
    }

    #[test]
    fn test_unified_query_request_builder() {
        let req = UnifiedQueryRequest::new(
            "SELECT * FROM users".to_string(),
            DatabaseType::MySQL,
        )
        .with_timeout(60)
        .with_limit(false, 0);

        assert_eq!(req.timeout_secs, 60);
        assert!(!req.apply_limit);
        assert_eq!(req.limit_value, 0);
    }

    #[test]
    fn test_unified_query_response_creation() {
        let response = UnifiedQueryResponse::new(
            "SELECT * FROM users".to_string(),
            "SELECT * FROM users LIMIT 1000".to_string(),
            DatabaseType::PostgreSQL,
            vec![serde_json::json!({"id": 1, "name": "Alice"})],
            150,
            true,
        );

        assert_eq!(response.row_count, 1);
        assert_eq!(response.execution_time_ms, 150);
        assert!(response.limit_applied);
        assert_eq!(response.database_type, DatabaseType::PostgreSQL);
    }
}
